File: go/internal/cli/doctor.go
package cli

import (
	"encoding/json"
	"errors"
	"fmt"
	"sort"

	"github.com/spf13/cobra"
)

func newDoctorCmd() *cobra.Command {
	var asJSON bool
	cmd := &cobra.Command{
		Use:   "doctor",
		Short: "Diagnose store, embedding backend, and configuration health",
		Long: "Actively probes the embedding backend (otherwise only contacted lazily on\n" +
			"first use), checks the store is reachable, guards against an embedding-\n" +
			"dimension mismatch, and reports the resolved configuration. Exits non-zero\n" +
			"if any check fails, so it doubles as a scriptable readiness gate.",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ctx := cmd.Context()
			store := storeFromCtx(ctx)
			cfg := cfgFromCtx(ctx)

			// Each check is a flat map {"name","ok", ...detail} — the same shape
			// the Python port's `doctor --json` emits, so a monitoring script
			// works against either port unchanged.
			var checks []map[string]any
			add := func(name string, ok bool, detail map[string]any) {
				c := map[string]any{"name": name, "ok": ok}
				for k, v := range detail {
					if k != "name" && k != "ok" && v != nil {
						c[k] = v
					}
				}
				checks = append(checks, c)
			}

			// 1. Resolved configuration.
			add("config", true, map[string]any{
				"store_path":     cfg.StorePath,
				"collection":     cfg.Collection,
				"embed_provider": cfg.EmbedProvider,
				"embed_model":    embedModelName(cfg),
				"embed_dim":      cfg.EmbedDim,
			})

			// 2. Store reachable.
			if n, err := store.Count(ctx); err != nil {
				add("store", false, map[string]any{"error": err.Error()})
			} else {
				add("store", true, map[string]any{"count": n})
			}

			// 3. Embedding backend reachability + dimension + latency. Strip the
			//    probe's own "ok" so it is not duplicated at both levels.
			probe := store.ProbeEmbedding(ctx)
			probeOK, _ := probe["ok"].(bool)
			detail := map[string]any{"model": embedModelName(cfg)}
			for k, v := range probe {
				if k != "ok" {
					detail[k] = v
				}
			}
			add("embedder", probeOK, detail)

			// 4. Embed-dim guardrail: the probe's dimension must match the
			//    configured dimension the backend was created with.
			if probeOK {
				if pd, ok := probe["dim"].(int); ok {
					add("embed_dim", pd == cfg.EmbedDim, map[string]any{
						"probe_dim": pd, "configured_dim": cfg.EmbedDim,
					})
				}
			}

			// 5. Audit journal reachable.
			add("journal", true, map[string]any{"enabled": store.Journal() != nil})

			ok := true
			for _, c := range checks {
				if b, _ := c["ok"].(bool); !b {
					ok = false
				}
			}

			if asJSON || fmtFromCtx(ctx) == FormatJSON {
				b, _ := json.MarshalIndent(map[string]any{"ok": ok, "checks": checks}, "", "  ")
				fmt.Println(string(b))
			} else {
				for _, c := range checks {
					mark := "[ok]"
					if b, _ := c["ok"].(bool); !b {
						mark = "[!!]"
					}
					// Sorted keys so the report is byte-stable run-to-run.
					keys := make([]string, 0, len(c))
					for k := range c {
						if k != "name" && k != "ok" {
							keys = append(keys, k)
						}
					}
					sort.Strings(keys)
					fmt.Printf("%s %-11s", mark, c["name"])
					for _, k := range keys {
						fmt.Printf(" %s=%v", k, c[k])
					}
					fmt.Println()
				}
				fmt.Println()
				if ok {
					fmt.Println("OK — ready")
				} else {
					fmt.Println("NOT READY — see failures above")
				}
			}

			if !ok {
				return errors.New("diagnostics failed")
			}
			return nil
		},
	}
	cmd.Flags().BoolVar(&asJSON, "json", false, "Emit the report as JSON")
	return cmd
}

File: ai_houkai/cli/commands/doctor.py
from __future__ import annotations

import json

import typer

from ai_houkai.cli import output as out


def doctor(
    ctx: typer.Context,
    as_json: bool = typer.Option(False, "--json", help="Emit the report as JSON"),
) -> None:
    """Diagnose store, embedding backend, and configuration health.

    Actively probes the embedding backend (which is otherwise only contacted
    lazily on the first write/recall), checks the store is reachable, guards
    against an embedding-dimension mismatch between the model and the stored
    vectors, and reports the resolved configuration. Exits non-zero if any
    check fails, so it doubles as a scriptable readiness gate.
    """
    store = ctx.obj["store"]
    checks: list[dict] = []

    def add(name: str, ok: bool, **detail: object) -> None:
        checks.append({"name": name, "ok": ok, **detail})

    # 1. Resolved configuration.
    add(
        "config", True,
        store_path=store.path,
        collection=store.collection_name,
        embedding_model=store.embedding_model,
        embedder=store.embedder_name,
    )

    # 2. Store reachable.
    count: int | None = None
    try:
        count = store.count()
        add("store", True, count=count)
    except Exception as exc:  # noqa: BLE001 — report, don't crash the diagnosis
        add("store", False, error=f"{type(exc).__name__}: {exc}")

    # 3. Embedding backend reachability + dimension + latency.
    probe = store.probe_embedding()
    add(
        "embedder", bool(probe.get("ok")),
        model=store.embedder_name,
        dim=probe.get("dim"),
        latency_ms=probe.get("latency_ms"),
        error=probe.get("error"),
    )

    # 4. Vector index readable, and its dimension matching the probe's — a
    #    mismatch makes cosine similarity silently meaningless, and an index
    #    that cannot be read at all makes every query answer with nothing
    #    while the collection still counts and lists its memories. Reading one
    #    row's embedding settles both. This used to swallow the read error and
    #    skip the check, so a collection that could not answer a single query
    #    was diagnosed "OK — ready".
    stored_dim: int | None = None
    if count:
        hint = "run `houkai rebuild-vectors` to rebuild it from the stored text"
        try:
            res = store.collection.get(limit=1, include=["embeddings"])
            embs = res.get("embeddings")
            if embs is None or len(embs) == 0:
                add("vector_index", False, memories=count,
                    error=f"the collection holds text but no vectors — {hint}")
            else:
                stored_dim = len(embs[0])
                add("vector_index", True, memories=count)
        except Exception as exc:  # noqa: BLE001 — report, don't crash the diagnosis
            add("vector_index", False, memories=count,
                error=f"{type(exc).__name__}: {exc} — {hint}")
    if stored_dim is not None and probe.get("ok"):
        add(
            "embed_dim", stored_dim == probe.get("dim"),
            probe_dim=probe.get("dim"), stored_dim=stored_dim,
        )

    # 5. Audit journal readable + freshness.
    jp = store.journal.path
    exists = jp.exists()
    detail: dict[str, object] = {
        "enabled": store._journal_enabled, "path": str(jp), "exists": exists,
    }
    if exists:
        st = jp.stat()
        detail["size_bytes"] = st.st_size
        detail["last_write_age"] = out.fmt_age(st.st_mtime)
    add("journal", True, **detail)

    ok = all(c["ok"] for c in checks)
    report = {"ok": ok, "checks": checks}

    if as_json:
        typer.echo(json.dumps(report, indent=2))
    else:
        for c in checks:
            mark = "[ok]" if c["ok"] else "[!!]"
            extras = " ".join(
                f"{k}={v}" for k, v in c.items()
                if k not in ("name", "ok") and v is not None
            )
            typer.echo(f"{mark} {c['name']:<11} {extras}")
        typer.echo("")
        typer.echo("OK — ready" if ok else "NOT READY — see failures above")

    if not ok:
        raise typer.Exit(1)
